cpp/log/cluster_state_controller-inl.h

```diff
@@ -0,0 +1,207 @@
+#ifndef CERT_TRANS_LOG_CLUSTER_STATE_CONTROLLER_INL_H_
+#define CERT_TRANS_LOG_CLUSTER_STATE_CONTROLLER_INL_H_
+
+#include <functional>
+
+#include "log/cluster_state_controller.h"
+#include "proto/ct.pb.h"
+
+#include "log/etcd_consistent_store-inl.h"
+
+namespace cert_trans {
+
+
+template <class Logged>
+ClusterStateController<Logged>::ClusterStateController(
+    ConsistentStore<Logged>* store, const MasterElection* election,
+    const int min_serving_nodes, const double min_serving_fraction)
+    : store_(CHECK_NOTNULL(store)),
+      election_(CHECK_NOTNULL(election)),
+      min_serving_nodes_(min_serving_nodes),
+      min_serving_fraction_(min_serving_fraction),
+      exiting_(false),
+      update_required_(false),
+      cluster_serving_sth_update_thread_(
+          std::bind(&ClusterStateController<Logged>::ClusterServingSTHUpdater,
+                    this)) {
+  CHECK_LE(1, min_serving_nodes_);
+  CHECK_LT(0, min_serving_fraction_);
+  store_->WatchClusterNodeStates(
+      std::bind(&ClusterStateController::OnClusterStateUpdated, this,
+                std::placeholders::_1));
+}
+
+
+template <class Logged>
+ClusterStateController<Logged>::~ClusterStateController() {
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    exiting_ = true;
+    update_required_cv_.notify_all();
```

Comment:
Argh, I missed the other ```notify_all()```, for the "release the lock" thing... Not incorrect, strictly speaking, just nicer...